cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,413 @@
+	maxRevocations  int
+	serialBatchSize int
+	raClient        revoker
+	mailer          mail.Mailer
+	emailSubject    string
+	emailTemplate   *template.Template
+	logger          log.Logger
+}
+
+// uncheckedBlockedKey represents a row in the blockedKeys table
+type uncheckedBlockedKey struct {
+	KeyHash   []byte
+	RevokedBy int64
+}
+
+func (bkr *badKeyRevoker) selectUncheckedKey() (uncheckedBlockedKey, error) {
+	var row uncheckedBlockedKey
+	err := bkr.dbMap.SelectOne(
+		&row,
+		"SELECT keyHash, revokedBy FROM blockedKeys WHERE extantCertificatesChecked = false",
+	)
+	return row, err
+}
+
+// unrevokedCertificate represents a yet to be revoked certificate
+type unrevokedCertificate struct {
+	Serial         string
+	DER            []byte
+	RegistrationID int64
+}
+
+// findUnrevoked looks for all unexpired, currently valid certificates which have a specific SPKI hash,
+// by looking first at the keyHashToSerial table and then the certificateStatus and certificates tables.
+// If the number of certificates it finds is larger than bkr.maxRevocations it'll error out.
+func (bkr *badKeyRevoker) findUnrevoked(unchecked uncheckedBlockedKey) ([]unrevokedCertificate, error) {
+	var unrevokedCerts []unrevokedCertificate
+	initialID := 0
+	for {
+		var batch []struct {
+			ID         int
+			CertSerial string
+		}
+		_, err := bkr.dbMap.Select(
+			&batch,
+			"SELECT id, certserial FROM keyHashToSerial WHERE keyHash = ? AND id > ? ORDER BY id LIMIT ?",
+			unchecked.KeyHash,
+			initialID,
+			bkr.serialBatchSize,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if len(batch) == 0 {
+			break
+		}
+		initialID = batch[len(batch)-1].ID
+		for _, serial := range batch {
+			var unrevokedCert unrevokedCertificate
+			err = bkr.dbMap.SelectOne(
+				&unrevokedCert,
+				`SELECT cs.serial, c.registrationID, c.der
+				FROM certificateStatus AS cs
+				JOIN certificates AS c
+				ON cs.serial = c.serial
+				WHERE cs.serial = ? AND cs.isExpired = false AND cs.status != ?`,
+				serial.CertSerial,
+				string(core.StatusRevoked),
+			)
+			if err != nil {
+				if db.IsNoRows(err) {
+					continue
+				}
+				return nil, err
+			}
+			unrevokedCerts = append(unrevokedCerts, unrevokedCert)
+		}
+	}
+	if len(unrevokedCerts) > bkr.maxRevocations {
+		return nil, fmt.Errorf("too many certificates to revoke associated with %x: got %d, max %d", unchecked.KeyHash, len(unrevokedCerts), bkr.maxRevocations)
+	}
+	return unrevokedCerts, nil
+}
+
+// markRowChecked updates a row in the blockedKeys table to mark a keyHash
+// as having been checked for extant unrevoked certificates.
+func (bkr *badKeyRevoker) markRowChecked(unchecked uncheckedBlockedKey) error {
+	_, err := bkr.dbMap.Exec("UPDATE blockedKeys SET extantCertificatesChecked = true WHERE keyHash = ?", unchecked.KeyHash)
+	return err
+}
+
+// resolveContacts builds a map of id -> email address. Due to backwards compat
+// the contact field in the database is represented as a slice, but only a single
```

Comment:
Why do you say only a single address should ever be present? We have a config field `maxContactsPerRegistration` and it appears to be set to 10 in prod. Is there another place where we're limiting to 1?